flink-table-store-connector/src/test/java/org/apache/flink/table/store/connector/FileStoreITCase.java

```diff
@@ -0,0 +1,197 @@
+
+    private static final RowType PARTITION_TYPE =
+            new RowType(Collections.singletonList(new RowType.RowField("p", new VarCharType())));
+
+    private static final RowType KEY_TYPE =
+            new RowType(Collections.singletonList(new RowType.RowField("k", new IntType())));
+
+    private static final RowType VALUE_TYPE =
+            new RowType(
+                    Arrays.asList(
+                            new RowType.RowField("v", new IntType()),
+                            new RowType.RowField("p", new VarCharType()),
+                            // rename key
+                            new RowType.RowField("_k", new IntType())));
+
+    @SuppressWarnings({"unchecked", "rawtypes"})
+    private static final DataStructureConverter<RowData, Row> CONVERTER =
+            (DataStructureConverter)
+                    DataStructureConverters.getConverter(
+                            TypeConversions.fromLogicalToDataType(VALUE_TYPE));
+
+    private static final int NUM_BUCKET = 3;
+
+    private static final List<RowData> SOURCE_DATA =
+            Arrays.asList(
+                    wrap(GenericRowData.of(0, StringData.fromString("p1"), 1)),
+                    wrap(GenericRowData.of(0, StringData.fromString("p1"), 2)),
+                    wrap(GenericRowData.of(5, StringData.fromString("p1"), 1)),
+                    wrap(GenericRowData.of(6, StringData.fromString("p2"), 1)),
+                    wrap(GenericRowData.of(3, StringData.fromString("p2"), 5)),
+                    wrap(GenericRowData.of(5, StringData.fromString("p2"), 1)));
+
+    private Configuration options;
+
+    private final boolean isBounded;
+
+    public FileStoreITCase(boolean isBounded) {
+        this.isBounded = isBounded;
+    }
+
+    @Parameterized.Parameters(name = "isBounded-{0}")
+    public static List<Boolean> getVarSeg() {
+        return Arrays.asList(true, false);
+    }
+
+    private static SerializableRowData wrap(RowData row) {
+        return new SerializableRowData(row, InternalSerializers.create(VALUE_TYPE));
+    }
+
+    @Before
+    public void before() throws IOException {
+        options = new Configuration();
+        options.set(BUCKET, NUM_BUCKET);
+        options.set(FILE_PATH, TEMPORARY_FOLDER.newFolder().toURI().toString());
+        options.set(FILE_FORMAT, "avro");
+    }
+
+    @Test
+    public void testPartitioned() throws Exception {
+        innerTest(true);
+    }
+
+    @Test
+    public void testNonPartitioned() throws Exception {
+        innerTest(false);
+    }
+
+    private void innerTest(boolean partitioned) throws Exception {
+        int[] partitions = partitioned ? new int[] {1} : new int[0];
+        int[] keys = new int[] {2};
+        InternalTypeInfo<RowData> typeInfo = InternalTypeInfo.of(VALUE_TYPE);
+
+        StreamExecutionEnvironment env = isBounded ? buildBatchEnv() : buildStreamEnv();
+        env.setParallelism(1); // no key by, must single parallelism
```

Comment:
If there is only one parallelism what's the point of setting the number of buckets to 3? Also this test fails to cover writing into multiple buckets and partitions at the same time.

Reply:
> If there is only one parallelism what's the point of setting the number of buckets to 3?

Why no point? A writer can accept records from multiple buckets.

Reply:
> Also this test fails to cover writing into multiple buckets and partitions at the same time.

The records are in multiple buckets and partitions. What is your point?

Reply:
The IT case does not cover the multiple writer instances situation.

> > Also this test fails to cover writing into multiple buckets and partitions at the same time.
> 
> The records are in multiple buckets and partitions. What is your point?

Reply:
> The IT case does not cover the multiple writer instances situation.

That is true, we can add a `key by` to support multiple parallelisms.